test/boost/logalloc_test.cc

```diff
@@ -831,6 +866,9 @@ SEASTAR_THREAD_TEST_CASE(background_reclaim) {
         while (thread_cputime_clock::now() < deadline) {
             thread::maybe_yield();
         }
```

Comment:
Your change means that we give up on testing the performance of background reclaim, and instead we only check that it does its job eventually.

With that change, there's no point in doing the 100ms sleeps anymore. The metric_change_monitor will already sleep as much as background reclaim needs. Additional hardcoded sleeps will only make the test slower.

Reply:
Yes. Options
 - remove those spins
 - remove the spins, add a perf-simple-query style test on semi-fixed isolated hardware
 - keep the spins, count the number of times we wait for the metrics, verify the total waits is below some empirically non-flaky number

Reply:
```diff
diff --git a/test/boost/logalloc_test.cc b/test/boost/logalloc_test.cc
index 6df11a2728..64285d5900 100644
--- a/test/boost/logalloc_test.cc
+++ b/test/boost/logalloc_test.cc
@@ -846,10 +846,15 @@ SEASTAR_THREAD_TEST_CASE(background_reclaim) {
     // Make sure the reclaimer catches up
     reactor_polls_metric.wait_for_change().get();
     reactor_tasks_pending_metric.wait_for_zero().get();
 
     std::vector<managed_bytes> std_allocs;
+    unsigned sleep_counts = 0;
+    auto count_sleeps_and_wait = [&] (future<> f) {
+        sleep_counts += !f.available();
+        f.get();
+    };
     size_t std_alloc_size = 1000000; // note that managed_bytes fragments these, even in std
     for (int i = 0; i < 50; ++i) {
         auto compacted_pre = logalloc::shard_tracker().statistics().memory_compacted;
         fmt::print("compacted {} items {} (pre)\n", compacted_pre, evictable_allocs.size());
         std_allocs.emplace_back(managed_bytes::initialized_later(), std_alloc_size);
@@ -864,14 +869,16 @@ SEASTAR_THREAD_TEST_CASE(background_reclaim) {
         // and causing test failures.
         auto deadline = thread_cputime_clock::now() + 100ms;
         while (thread_cputime_clock::now() < deadline) {
             thread::maybe_yield();
         }
-        // Make sure the reclaimer catches up
+        // Make sure the reclaimer catches up, but count how many times we had to do it.
         reactor_polls_metric.wait_for_change().get();
-        reactor_tasks_pending_metric.wait_for_zero().get();
+        count_sleeps_and_wait(reactor_tasks_pending_metric.wait_for_zero());
     }
+    // Allow the reclaimer not too keep up due to the machine being overloaded, but not all the time.
+    BOOST_REQUIRE_LE(sleep_counts, 15);
 }
 
 inline
 bool is_aligned(void* ptr, size_t alignment) {
     return uintptr_t(ptr) % alignment == 0;
```

Reply:
testing this ^

Reply:
```
test/boost/logalloc_test.cc(879): fatal error: in "background_reclaim": critical check sleep_counts <= 15 has failed [50 > 15]
```
??

Reply:
now this:

```diff
diff --git a/test/boost/logalloc_test.cc b/test/boost/logalloc_test.cc
index 6df11a2728..499cb05292 100644
--- a/test/boost/logalloc_test.cc
+++ b/test/boost/logalloc_test.cc
@@ -846,10 +846,15 @@ SEASTAR_THREAD_TEST_CASE(background_reclaim) {
     // Make sure the reclaimer catches up
     reactor_polls_metric.wait_for_change().get();
     reactor_tasks_pending_metric.wait_for_zero().get();
 
     std::vector<managed_bytes> std_allocs;
+    unsigned sleep_counts = 0;
+    auto count_sleeps_and_wait = [&] (future<> f) {
+        sleep_counts += !f.available();
+        f.get();
+    };
     size_t std_alloc_size = 1000000; // note that managed_bytes fragments these, even in std
     for (int i = 0; i < 50; ++i) {
         auto compacted_pre = logalloc::shard_tracker().statistics().memory_compacted;
         fmt::print("compacted {} items {} (pre)\n", compacted_pre, evictable_allocs.size());
         std_allocs.emplace_back(managed_bytes::initialized_later(), std_alloc_size);
@@ -864,14 +869,18 @@ SEASTAR_THREAD_TEST_CASE(background_reclaim) {
         // and causing test failures.
         auto deadline = thread_cputime_clock::now() + 100ms;
         while (thread_cputime_clock::now() < deadline) {
             thread::maybe_yield();
         }
-        // Make sure the reclaimer catches up
-        reactor_polls_metric.wait_for_change().get();
-        reactor_tasks_pending_metric.wait_for_zero().get();
+        if (memory::stats().free_memory() < 30'000'000) {
+            // Make sure the reclaimer catches up, but count how many times we had to do it.
+            reactor_polls_metric.wait_for_change().get();
+            count_sleeps_and_wait(reactor_tasks_pending_metric.wait_for_zero());
+        }
     }
+    // Allow the reclaimer not too keep up due to the machine being overloaded, but not all the time.
+    BOOST_REQUIRE_LE(sleep_counts, 15);
 }
 
 inline
 bool is_aligned(void* ptr, size_t alignment) {
     return uintptr_t(ptr) % alignment == 0;
```

Reply:
> now this:

We start with 60e6 bytes of free memory. The test allocates 1e6 bytes 50 times.

In order to reach your `memory::stats().free_memory() < 30'000'000` condition, we need at least 30 loops. But after this happens, the `wait_for_zero` sleep will let the background reclaimer reset the free memory back to 60e6. To hit your condition again, we would need another 30 loops. But we only have 20 loops left. So we don't even have a chance to fail 2 times, let alone 15 times, regardless of performance. Not a very useful test.

Reply:
It won't recover everything so easily, at the beginning it has few shares and it's competing with a spin loop, not sleep.

I can try to reduce the 15 to something smaller, but we risk bringing the flakiness back. In the end a performance test like this can't be reliable in an overprovisioned environment.

Reply:
```diff
diff --git a/test/boost/logalloc_test.cc b/test/boost/logalloc_test.cc
index 499cb05292..b53c55eaa5 100644
--- a/test/boost/logalloc_test.cc
+++ b/test/boost/logalloc_test.cc
@@ -869,18 +869,18 @@ SEASTAR_THREAD_TEST_CASE(background_reclaim) {
         // and causing test failures.
         auto deadline = thread_cputime_clock::now() + 100ms;
         while (thread_cputime_clock::now() < deadline) {
             thread::maybe_yield();
         }
-        if (memory::stats().free_memory() < 30'000'000) {
+        if (memory::stats().free_memory() < 50'000'000) {
             // Make sure the reclaimer catches up, but count how many times we had to do it.
             reactor_polls_metric.wait_for_change().get();
             count_sleeps_and_wait(reactor_tasks_pending_metric.wait_for_zero());
         }
     }
     // Allow the reclaimer not too keep up due to the machine being overloaded, but not all the time.
-    BOOST_REQUIRE_LE(sleep_counts, 15);
+    BOOST_REQUIRE_LE(sleep_counts, 10);
 }
 
 inline
 bool is_aligned(void* ptr, size_t alignment) {
     return uintptr_t(ptr) % alignment == 0;
```

Reply:
> It won't recover everything so easily, at the beginning it has few shares and it's competing with a spin loop, not sleep.

No. At the beginning of the test and every time we fall under 30'000'000, you are calling `wait_for_zero`, which gives background_reclaim all the time it wants (and it wants to bring free memory back to to 60'000'000).

Reply:
wait_for_zero() only waits for anything if the reclaimer is active. But I guess as soon as the timer fires it will be active until it catches up.